server/queue/queue-pulsar/src/main/scala/org/apache/james/queue/pulsar/PulsarMailQueue.scala

```diff
@@ -0,0 +1,600 @@
+    streams.committableSource(scheduledConsumer)
+      .toMat(requeueMessage)(Keep.left) //use of toMat to keep reference of Control which would be lost by direct usage of flatMapConcat
+
+  private val filterStage: ActorRef = system.actorOf(FilterStage.props)
+
+  private val counter: Sink[Any, Future[Int]] =
+    Sink.fold[Int, Any](0)((acc, _) => acc + 1)
+
+  private val dequeueFlow: RunnableGraph[(Control, Publisher[MailQueueItem])] = {
+    implicit val timeout: Timeout = Timeout(1, TimeUnit.SECONDS)
+    streams.committableSource(consumer)
+      .map(message =>
+        (Json.fromJson[MailMetadata](Json.parse(message.message.value)).get,
+          message)
+      ).ask[(Option[MailMetadata], Option[MimeMessagePartsId], CommittableMessage[String])](filterStage)
+      .flatMapConcat {
+        case (None, Some(partsId), committableMessage) =>
+          committableMessage.ack()
+          deleteMimeMessage(partsId)
+            .flatMapConcat(_ => Source.empty)
+        case (Some(metadata), _, committableMessage) =>
+          val partsId = metadata.partsId
+          Source
+            .fromPublisher(readMimeMessage(partsId))
+            .map(message => (readMail(metadata, message), partsId, committableMessage))
+      }.map { case (mail, partsId, message) => new PulsarMailQueueItem(mail, partsId, message) }
+      .map(mailQueueItemDecoratorFactory.decorate(_, name))
+
+      .alsoTo(counter)
+      // akka streams virtual publisher handles a subscription timeout to the
+      // exposed publisher which will terminate the stream if the timeout is not
+      // honored. To do so, the akka stream implementation actually subscribes to
+      // the publisher.
+      // asPublisher thus requires either :
+      // * fanout=true
+      // * or to disable the subscription timeout mechanism
+      // see akka.stream.impl.VirtualPublisher.onSubscriptionTimeout
+
+      .via(debugLogger("dequeueFlow"))
+      .toMat(Sink.asPublisher[MailQueue.MailQueueItem](true).withAttributes(Attributes.inputBuffer(initial = 1, max = 1)))(Keep.both)
+  }
+
+  class PulsarMailQueueItem(mail: Mail, partsId: MimeMessagePartsId, message: CommittableMessage[String]) extends MailQueueItem {
+    override val getMail: Mail = mail
+
+    override def done(success: Boolean): Unit = {
+      if (success) {
+        dequeueMetrics.increment()
+        Await.ready(message.ack(cumulative = false), awaitTimeout)
+        deleteMimeMessage(partsId).run()
+      } else {
+        Await.ready(message.nack(), awaitTimeout)
+      }
+    }
+  }
+
+  /**
+   * For now, filterFlow always rereads the whole topic from the start by using a random subscription name.
+   * Filters are never removed from the topic.
+   * This means that the FilterStage will get slower to start as the number of filter increases, it will also consume
+   * an increasing amount of RAM until the first mail is processed which will invalidate and purge the expired filters.
```

Comment:
Given we could get emails with arbitrary delays, invalidating filters likely lead to bugs, as the underlying emails can be out of order.

Maybe it is safer to get rid of this optimisation. Correct?

Reply:
I think this is the same as the issue you raised on clear. The behavior for scheduled messages is not explicitly defined by the contracts. 
Removing the optimization means introducing an ever increasing list of filters which is akin to a memory leak. What we could look into is purging all scheduled messages that have been scheduled before the call to remove(), it requires a bit of rework but should be doable.